### PR TITLE
Windows: Builder case insensitive env

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>

@duglin - Addresses your comment https://github.com/docker/docker/pull/28667#issuecomment-262026383. Specifically the builder piece of the whole puzzle. I'll look at -e for `docker run` in a follow-up PR to keep things a little more succinct and by area.

This PR ensures that on Windows, environment variables in the builder (both as an `ENV` command and for variable substitution) are treated case-insensitive.

@thaJeztah @vieux As for https://github.com/docker/docker/pull/28667, this probably also should make it into 1.13.

### CODE REVIEW

**Comment by thaJeztah** [comment]
@jhowardmsft one question; is there a reason for picking _lowercase_ as the canonical? Is it more common on Windows to have env vars in lowercase (Linux is overall uppercase)?

**Comment by jhowardmsft** [comment]
No reason at all. If I used upper, someone would question it the other way 😇

**Comment by thaJeztah** [comment]
Just wondered what most "natural" was, I.e., least chance of it changing case, and least surprising to users if they specified a var

**Comment by jhowardmsft** [comment]
I honestly don't know. I've seen both

File: builder/dockerfile/shell_parser_test.go
Comment (duglin) [inline builder/dockerfile/shell_parser_test.go]: Can we add a check to make sure that words[0] is W, A or U and not something else?

**Review by duglin — APPROVED** [review]
One minor ask, otherwise LGTM

**Comment by jhowardmsft** [comment]
@thaJeztah Quick survey looks like most of the built-in vars are caps based on hangover days from DOS. Some mixed case and a handful of lowercase. I'll update to upper, and also address the comment from @duglin about the addition check to make sure W/A/U is valid shortly.

**Comment by jhowardmsft** [comment]
Updated to address comments.

**Comment by duglin** [comment]
still LGTM once janky is done

**Comment by thaJeztah** [comment]
Thanks for checking @jhowardmsft, 👍 for changing to uppercase. Guess we need to do the same for `docker run`?

**Comment by jhowardmsft** [comment]
@thaJeztah Yes, I have another PR to deal with deduping on docker run and uppercasing. Will be submitting that today, I'm just finishing up the integration test for it. There's various pieces of the whole puzzle here as I mentioned in the description of this PR.

File: builder/dockerfile/dispatchers.go
Comment (tonistiigi) [inline builder/dockerfile/dispatchers.go]: Why do we need to force a case in commit? Isn't case insensitive comparison enough?

**Comment by jhowardmsft** [comment]
@thaJeztah docker run (and docker create or anything that populates `config.Env` in the runconfig) PR submitted - #28770.

File: builder/dockerfile/dispatchers.go
Comment (jhowardmsft) [inline builder/dockerfile/dispatchers.go]: It means the comment written by the builder (which is used for hash generation to determine if a layer has changed) is consistently uppercase.

Comment (tonistiigi) [inline builder/dockerfile/dispatchers.go]: It does change the actual env key also, right? Not only the comment. If this is true then it can break programs that don't just use `getenv` but look at the keys directly. The fix should actually be in https://github.com/docker/docker/pull/28725/files#diff-77aa50fa42bcdf6c3bedc962a396b6eaR84 then.

Comment (jhowardmsft) [inline builder/dockerfile/dispatchers.go]: Yes, it does. But I think this is still in the right place due to https://github.com/docker/docker/pull/28770 which will ensure the runconfig is in uppercase and deduped too.

Comment (tonistiigi) [inline builder/dockerfile/dispatchers.go]: I'd say #28770 has the same problem. It should only do the case insensitive comparison and leave the last key in its original form. Also see, https://github.com/docker/docker/blob/master/daemon/commit.go#L46-L56 for more unhandled cases.

Comment (jhowardmsft) [inline builder/dockerfile/dispatchers.go]: > It should only do the case insensitive comparison and leave the last key in its original form

@tonistiigi I disagree. As I've said environment variables are completely case insensitive on Windows. It's better to be consistent here and have docker manage the environment variables in single case stored in the config and where they're handled.

```
C:\> $env:FoObAr="something"
C:\> $env:foobar
something
C:\> $env:FOOBAR
something
```

But I do agree that commit.go also needs changing. Will update that.

Comment (tonistiigi) [inline builder/dockerfile/dispatchers.go]: They are not case insensitive in a way that you can query all environment variables and you will see their keys in the original form. Not normalized to any specific form or uppercased. This means that although `$env:foobar` and `$env:FOOBAR` have the same value an application can specifically check if `foobar` is in the environment without matching the uppercase form.

Comment (jhowardmsft) [inline builder/dockerfile/dispatchers.go]: @tonistiigi I don't have a strong enough opinion to justify it either way, but updated to just do the comparison.

**Comment by tonistiigi** [comment]
LGTM

nit: could probably save some lines and cpu cycles by doing
```
var equalEnv = func(s1, s2 string)bool {return s1 == s2}
func init() {
  if windows {
   equalEnv = strings.EqualFold
  }
}
```
but not needed if you prefer converting.

**Comment by thaJeztah** [comment]
ping @duglin PTAL

**Comment by jhowardmsft** [comment]
@duglin - Hey Doug - just think @vieux wants another quick review after the latest revision so we can get this into 1.13. Any chance you have time to review/LGTM? Thanks :smile:

**Comment by duglin** [comment]
yep - still LGTM

**Pull request merged**
